### ISSUE TITLE
How to do Inline fragments in TypeSafe client

### ISSUE THREAD

**Issue by phillip-kruger** [smallrye/smallrye-graphql#867]
see https://graphql.org/learn/queries/#inline-fragments

Now sure if we can do this yet ? 

@t1 ?

/cc @xstefank

**Comment by t1** [comment]
I understand why Fragments are being used for, but I don't see how this would apply to the typesafe client. Do you have some code in mind to explain it?

**Comment by phillip-kruger** [comment]
For example, using the GitHub API:

```
{
  search(query: "...bla bla bla", type: ISSUE, last: 100) {
    edges {
      node {
        ... on PullRequest {
          number
          title
          repository {
            nameWithOwner
          }
          createdAt
          mergedAt
          url
          changedFiles
          additions
          deletions
          labels(last: 100) {
            edges {
              node {
                name
                color
              }
            }
          }
        }
      }
    }
  }
}
```

So Node has no fields, you have to select fields on a specific concrete implementation

**Comment by phillip-kruger** [comment]
https://docs.github.com/en/graphql/overview/explorer

**Comment by t1** [comment]
Yeah, I meant Java code ;-) What's the use-case in the Java code using the typesafe client?

**Comment by phillip-kruger** [comment]
You mean in this case you should just use the Dynamic client ? I don't think that is ok. We should be able to model this

**Comment by t1** [comment]
No, when you request the node with the typesafe client, you'd directly request a `PullRequest`; while wrapping it with the edges is still necessary. I'm not sure how to map the if-this-type-then-these-fields-if-that-type-then-those-fields logic to Java. Maybe we'd need some sort of generic `Fragment` types with 2,3,4,...,n type parameters, e.g. `Fragment2<Issue, PullRequest>`.

**Comment by phillip-kruger** [comment]
I was thinking to just allow a `@On` or `@From` annotation, that you can add to a `PullRequest` so

```
@On("node")
public class PullRequest {
}
```

or allow an interface (in this case `Node`) that can be annotated with `@On` or something

**Comment by phillip-kruger** [comment]
```
public class PullRequest implements Node{
}

@On
public interface Node{
}

```

**Comment by t1** [comment]
Hmmm... we could check every return type to see if it has subtypes/implementations. If so, we could request `__typename` and all the possible fragments. Probably the hardest part will be to map the returned JSON to the specific types. JSON-B doesn't do that OOTB.

**Comment by phillip-kruger** [comment]
That is when you use the interface right ? The annotation on the class can still work ?

**Comment by xstefank** [comment]
so after discussing it with @jmartisk it seems that even the dynamic client isn't able to do fragments.

**Comment by phillip-kruger** [comment]
What ?!? Even if passing in a String ?

**Comment by xstefank** [comment]
@phillip-kruger no, passing a String works. Sorry, I meant embedded directly in the API.

**Comment by phillip-kruger** [comment]
OK cool.

**Comment by jmartisk** [comment]
Yeah, the DSL doesn't have that yet. I reported https://github.com/smallrye/smallrye-graphql/issues/871 so we don't forget about it

**Comment by t1** [comment]
If I understand this correctly, Fragments only make sense when you have Unions (or at least Interfaces), right?

**Comment by jmartisk** [comment]
They can also be used to prevent having to repeat a large chunk of fields several times, like in the example here
https://graphql.org/learn/queries/#fragments 
but this is chiefly a problem when writing queries by hand, not so much when building queries programmatically
